checks/elasticsearch.go

```diff
@@ -0,0 +1,93 @@
+package checks
+
+import (
+	//gocontext "context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/flanksource/canary-checker/api/context"
+	"github.com/flanksource/canary-checker/api/external"
+
+	v1 "github.com/flanksource/canary-checker/api/v1"
+	"github.com/flanksource/canary-checker/pkg"
+
+	"github.com/elastic/go-elasticsearch/v8"
+)
+
+type ElasticsearchChecker struct{}
+
+func (c *ElasticsearchChecker) Type() string {
+	return "elasticsearch"
+}
+
+func (c *ElasticsearchChecker) Run(ctx *context.Context) pkg.Results {
+	var results pkg.Results
+	for _, conf := range ctx.Canary.Spec.Elasticsearch {
+		results = append(results, c.Check(ctx, conf)...)
+	}
+	return results
+}
+
+func (c *ElasticsearchChecker) Check(ctx *context.Context, extConfig external.Check) pkg.Results {
+	check := extConfig.(v1.ElasticsearchCheck)
+	result := pkg.Success(check, ctx.Canary)
+	var results pkg.Results
+	results = append(results, result)
+
+	cfg := elasticsearch.Config{
+		Addresses: []string{check.GetEndpoint()},
+		Username:  check.Auth.GetUsername(),
+		Password:  check.Auth.GetPassword(),
+	}
+
+	es, err := elasticsearch.NewClient(cfg)
+	if err != nil {
+		return results.ErrorMessage(err)
+	}
+
+	body := strings.NewReader(check.Query)
+
+	res, err := es.Search(
+		es.Search.WithIndex(check.Index),
+		es.Search.WithBody(body),
+	)
+
+	if err != nil {
+		return results.ErrorMessage(err)
+	}
+
+	if res.IsError() {
+		var e map[string]interface{}
+		if err := json.NewDecoder(res.Body).Decode(&e); err != nil {
+			return results.ErrorMessage(
+				fmt.Errorf("Error parsing the response body: %s", err),
+			)
+		} else {
+			return results.ErrorMessage(fmt.Errorf("Error from elasticsearch.[%s]: %v, %v",
+				res.Status(),
+				e["error"].(map[string]interface{})["type"],
+				e["error"].(map[string]interface{})["reason"],
+			))
+		}
+	}
+
+	// We are closing the body after error as the Body object is not set in case of error
+	// leading to nil pointer errors
+	defer res.Body.Close()
```

Comment:
I would add this just after we create the res... since in case of res.IsError() we would be returning with a close body.

Reply:
As mentioned in the comment, if we put this before res.IsError and there is an error, we will get a nil pointer dereference error ... previously it was being used when the res object was being created but it was giving runtime errors in case of a failure scenario

Reply:
Ah I see